### PR TITLE
[PLATFORM-128] Smaller releaseables

### PR DESCRIPTION
**Pull request by mondoreale**
Besides a fancy name this PR is about lazy-loading the password library – `zxcvbn`.

**The idea**
I'm using dynamic `import` to load the library on demand.

Initially I had an idea to utilize `React.lazy` but `zxcvbn` is also being used in one of the utils so, this way seems to be more flexible.

I defined a toy called `PasswordStrength`. It uses render props to pass the calculated strength to underlying structures. It's responsible for lazy-loading `zxcvbn`.

Guys, let me know your thoughts!

---

screenshot 2018-11-28 11 21 50

### CODE REVIEW

**Comment by mondoreale** [comment]
It's greeen! 💚

**Comment by juhah** [comment]
Couldn't really find a way to test this properly... how you see that the async loading is happening?

On a side note, I found one issue with the login and signup form while trying this: https://streamr.atlassian.net/browse/PLATFORM-262

File: app/src/shared/components/PasswordStrength/index.jsx [hunk 1]
```diff
@@ -0,0 +1,58 @@
...
+        if (fakeStrength != null) {
+            return fakeStrength
+        }
+
+        return [0, 1, 1, 2, 2][(await zxcvbn())(value || '').score]
+    }
+
+    measure() {
+        this.getStrength().then((strength) => {
+            this.setState({
```
Comment (timoxley) [inline app/src/shared/components/PasswordStrength/index.jsx]: needs to check for unmount

File: app/package.json [hunk 2]
```diff
@@ -84,6 +84,7 @@
     "babel-eslint": "^8.2.2",
     "babel-jest": "^23.6.0",
     "babel-loader": "^7.1.2",
+    "babel-plugin-dynamic-import-node": "^2.2.0",
```
Comment (timoxley) [inline app/package.json]: IIUC this wouldn't be required if we let webpack handle imports rather than babel, right?

File: app/.babelrc [hunk 3]
```diff
@@ -4,5 +4,12 @@
   ],
   "plugins" : [
     "transform-flow-comments"
-  ]
+  ],
+  "env": {
+    "test": {
+      "plugins": [
+        "dynamic-import-node"
```
Comment (timoxley) [inline app/.babelrc]: why only in test?

File: app/test/unit/components/PasswordStrength/index.test.jsx [hunk 4]
```diff
@@ -0,0 +1,90 @@
...
+
+describe(PasswordStrength.name, () => {
+    const sandbox = sinon.createSandbox()
+
+    const getStrength = async (props) => (
+        mount(<PasswordStrength {...props} />).instance().getStrength()
+    )
+
+    const stubZxcvbn = (score) => {
+        sandbox.stub(zxcvbn, 'default').callsFake(() => Promise.resolve(() => ({
```
Comment (timoxley) [inline app/test/unit/components/PasswordStrength/index.test.jsx]: why not let it actually use the module? e.g. if the `zxcvbn` module ever changes and breaks our code, this test wouldn't catch that. This page of tests are kinda just testing that this array is in this shape: `[0, 1, 1, 2, 2]`, which is just configuration details. i.e. I don't think tests should be breaking on configuration changes. IMO find some examples that meet these strength criteria. If the module API changes or it starts considering weak passwords as strong, then these tests would catch that. That seems valuable. Maybe?

Comment (mondoreale) [reply app/test/unit/components/PasswordStrength/index.test.jsx]: Yep, you're right. Will change that.

File: app/.babelrc (hunk 3 above)
Comment (mondoreale) [reply app/.babelrc]: Tried it for all environments and.. code splitting stopped working. Apparently this plugin does not let webpack do its job. Looks like with it `import()` is more like `require()`.

Comment (mondoreale) [reply app/.babelrc]: And I added it for tests cause I aimed pretty much at `node` itself, not our webpack bundle.

File: app/package.json (hunk 2 above)
Comment (mondoreale) [reply app/package.json]: I'm not sure about that. All I know is that `node` does not recognize dynamic `import` syntax (`npm test`):

screenshot 2018-12-03 12 04 19

…and adding `dynamic-import-node` fixes it.

File: app/src/shared/errors/CancelledPromiseError/index.js [hunk 5]
```diff
@@ -0,0 +1,5 @@
+// @flow
+
+class CancelledPromiseError {}
```
Comment (juhah) [inline app/src/shared/errors/CancelledPromiseError/index.js]: Missing `extends Error`?

File: app/src/utils/makeCancelable.js [hunk 6]
```diff
@@ -0,0 +1,35 @@
...
+    promise: Promise<any>,
+    cancel: () => void,
+}
+
+/*
+ * NOTE: Took it from https://github.com/facebook/react/issues/5465#issuecomment-157888325 and
+ *       revamped it a little bit.
+ */
+
+const makeCancelable = (promise: Promise<any>) => {
```
Comment (juhah) [inline app/src/utils/makeCancelable.js]: You could call this file `promise.js` and add a comment that the purpose is to make any promise cancelable. Just to add some context :)

File: app/webpack.config.js [hunk 7]
```diff
@@ -25,7 +25,7 @@ const root = path.resolve(__dirname)
 const gitRevisionPlugin = new GitRevisionPlugin({
     gitWorkTree: path.resolve(root, '..'),
 })
-const publicPath = process.env.PLATFORM_BASE_PATH || '/'
+const publicPath = `${process.env.PLATFORM_BASE_PATH || ''}/`.replace(/\/+$/, '/')
```
Comment (timoxley) [inline app/webpack.config.js]: Worthwhile adding a comment for what this does.

File: app/src/shared/components/PasswordStrength/index.jsx [hunk 8]
```diff
@@ -0,0 +1,86 @@
...
+            this.getZxcvbn = makeCancelable(zxcvbn())
+            this.getZxcvbn.promise.then((measurer) => {
+                this.setState({
+                    measurer,
+                }, this.measure)
+            }, (error) => {
+                if (!(error instanceof CancelledPromiseError)) {
+                    throw error
+                }
+            })
```
Comment (timoxley) [inline app/src/shared/components/PasswordStrength/index.jsx]: Making a cancellable promise is cool… arguably more "correct" but it seems like a lot of mucking around compared to just setting a flag in `componentWillUnmount`:

```js
componentWillUnmount() {
    this._unmounted = true
}

loadMeasurer() {
    const { enabled, value } = this.props
    if (!enabled || !value) { return }
    const measurer = await zxcvbn()
    if (this._unmounted) { return }
    this.setState({ measurer }, this.measure)
} 
```

Cancelling is the right approach if there's something that can actually be cancelled, e.g. an expensive task, not sure it's the right option here.

File: app/src/shared/components/PasswordStrength/index.jsx [hunk 9]
```diff
@@ -0,0 +1,86 @@
...
+
+class PasswordStrength extends PureComponent<Props, State> {
+    state = {
+        measurer: null,
+        strength: -1,
+    }
+
+    componentDidMount() {
+        this.loadMeasurer()
+    }
```
Comment (timoxley) [inline app/src/shared/components/PasswordStrength/index.jsx]: Also when when this mounts, it goes on a real journey:

1. `componentDidMount` calls `loadMeasurer`
2. `loadMeasurer` calls `setState({measurer})` which causes a `render` + `componentDidUpdate`
3. `componentDidUpdate` calls `measure`
4. `measure` calls `setState({ strength })` which triggers a `render` + `componentDidUpdate`, which triggers another `measure` (is fine, no infinite loop prevented by `PureComponent`)
5. Second arg to `loadMeasurer` calls `measure` again.

It's fine, not a performance concern, just thinking perhaps this thing could be overall simpler? e.g.

```js
class PasswordStrength extends PureComponent<Props, State> {
    state = {
        strength: -1,
    }

    componentDidMount() {
        this.measure()
    }

    componentDidUpdate() {
        this.measure()
    }

    componentWillUnmount() {
        this.unmounted = true
    }

    unmounted: boolean

    measure = async () => {
        const { enabled, value } = this.props
        if (!enabled || !value) { return }

        const measurer = await zxcvbn()
        if (this.unmounted) { return }
        this.setState({
            strength: [0, 1, 1, 2, 2][measurer(value || '').score],
        })
    }

    render() {
        const { children, enabled, value } = this.props
        const strength = (enabled && value) ? this.state.strength : -1
        return children ? children(strength) : null
    }
}
```
Your call, just a suggestion.

**Comment by mondoreale** [comment]
@timoxley @juhah thank you guys for having another look at it! I pulled in @timoxley's ideas about rendering rounds and simplifications for "unmounted vs async/await". It's better now! 💪 

I'll wait for a green CI build and merge. Cheers!

**Pull request merged**